### PR TITLE
Support 'list' type in `output_wrapper_field_path`

### PR DESCRIPTION
**Pull request by brycahta**
Issue #, if available: https://github.com/aws-controllers-k8s/community/issues/1285

Description of changes:
* adds support for list types to `output_wrapper_field_path`
* cleans up logic in `setResourceReadMany`
* updates EC2 unit tests in model and code packages

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by jaypipes — CHANGES_REQUESTED** [review]
@brycahta I think this code might be getting a little too icky. I'm thinking perhaps it would be better if we instead took an approach of generating a separate `findSourceShape()` function that would return a singular shape that we wish to set the CR's fields from and have `SetResource()` only contain code that loops over a singular Shape's member fields.

In other words, we'd want to **generate** Go code that looks something like this:

```handlebars
// findSourceShape{{ $OpType }} returns the Shape that contains the fields
// we want to set on our CR when given an Output shape for the {{ $OpType }}
// Operation
func (rm *resourceManager) findSourceShape{{ $OpType }}(
	resp *svcsdk.{{$crd.Ops.$OpType.OutputRef.Shape.ShapeName }},
) (*svcsdk.{{ $crd.GetSingularStructNameForOp $OpType }}, error) {
	{{ GoCodeFindSingularSource $crd $OpType }}
}
```

We would need a `pkg/model.CRD.GetSingularStructNameForOp` method that would return the **name** of the `sdksvc` struct that contains the fields that we want to set a CR's field values from.

It might look something like this:

```go
// GetSingularStructNameForOp returns the name of the struct
// in the svcsdk package that contains fields that we want to set
// a CR's field values from.
//
// Any "wrapper" fields in an output shape will be popped as needed.
//
// For List/ReadMany operations, for example, this method will
// return the name of the struct containing a single resource
// representation within the list.
func (crd *CRD) GetSingularStructNameForOp(
	opType mode.OpType,
) string {
	// .. code here would pop any wrapper fields and
	// pop any list container fields, returning just the
	// name of the struct that represents a single resource
	// within the Operation's Output shape...
}
```

We would need to implement a new Go-code-outputting function in `pkg/generate/code` called `FindSingularSource()` that would produce the implementation of the `findSourceShape` resource manager method referenced in the template above.

Something like this...

```go
// FindSingularSource produces the Go code that accepts an Output shape
// from the response of an operation and returns the struct containing the
// singular resource representation that we will set a CR's fields from.
//
// For List/ReadMany operations, the code produced from this function might
// look something like this (example from VPC resource in EC2 controller):
//
// if len(resp.Vpcs) == 1 {
//     return resp.Vpcs[0]
// }
// return nil
//
// For other operations such as EC2's RunInstances output response, the
// code might look like this instead:
//
// if len(resp.Reservations) == 1 {
//     if len(resp.Reservations[0].Instances) == 1 {
//         return resp.Reservations[0].Instances[0]
//     }
// }
// return nil
//
// For ReadOne operations, where the response Output shape contain a
// "wrapper" field that contains the singular resource fields,
// the code returned from this function might look like this,
// such as the RDS CreateGlobalCluster Output shape in this example (note
// that "GlobalCluster" is the "wrapper" field
//
// if resp.GlobalCluster != nil {
//     return resp.GlobalCluster
// }
// return nil
func FindSingularSource(
	r *model.CRD,
	opType model.OpType,
	respVarName string,
	indentLevel int,
) string {
	// Code returned here would contain all the logic to "find"
	// a struct within a response output struct that
	// represents the singular resource. This struct
	// might a wrapper field or an element field in a
	// list container field...
}
```

We could then update our [`sdkFind` template](https://github.com/aws-controllers-k8s/code-generator/blob/de8d784c26215618088548f83d55dbf9ec2df6b1/templates/pkg/resource/sdk_find_read_many.go.tpl#L40-L46) in the code-generator from this:

```handlebars

	// Merge in the information we read from the API call above to the copy of
	// the original Kubernetes object we passed to the function
	ko := r.ko.DeepCopy()
{{- if $hookCode := Hook .CRD "sdk_read_many_pre_set_output" }}
{{ $hookCode }}
{{- end }}
{{ GoCodeSetReadManyOutput .CRD "resp" "ko" 1 }}
```

to something like this:

```handlebars
	// Pop the singular resource element from our response shape
	respSingle := rm.findSourceShapeReadMany(resp)
	// Merge in the information we read from the API call above to the copy of
	// the original Kubernetes object we passed to the function
	ko := r.ko.DeepCopy()
{{- if $hookCode := Hook .CRD "sdk_read_many_pre_set_output" }}
{{ $hookCode }}
{{- end }}
{{ GoCodeSetReadManyOutput .CRD "respSingle" "ko" 1 }}
```

File: pkg/generate/code/set_resource.go [hunk 1]
```diff
@@ -116,9 +113,28 @@ func SetResource(
...
 	wrapperFieldPath := r.GetOutputWrapperFieldPath(op)
-	if wrapperFieldPath != nil {
+	if op == r.Ops.ReadMany {
+		return setResourceReadMany(
+			cfg, r,
+			op, sourceVarName, targetVarName, indentLevel,
+		)
+	} else if wrapperFieldPath != nil {
+		// fieldpath api requires fully-qualified path
+		qwfp := fieldpath.FromString(op.OutputRef.ShapeName + "." + *wrapperFieldPath)
```
[the remark this answers is not in the document]
Comment (brycahta) [reply pkg/generate/code/set_resource.go:128]: qualified

File: pkg/generate/code/set_resource.go [hunk 2]
```diff
@@ -116,9 +113,28 @@ func SetResource(
...
+	// then call setResourceReadMany to generate for-range loops.
+	// Output shape will be a list for ReadMany operations or if
+	// designated via output wrapper config.
 	wrapperFieldPath := r.GetOutputWrapperFieldPath(op)
-	if wrapperFieldPath != nil {
+	if op == r.Ops.ReadMany {
+		return setResourceReadMany(
+			cfg, r,
+			op, sourceVarName, targetVarName, indentLevel,
+		)
```
[the remark this answers is not in the document]
Comment (brycahta) [reply pkg/generate/code/set_resource.go:125]: Yeah, I wanted to co-locate the `setResourceReadMany` branching logic for readability and to hopefully help with future refactoring

**Comment by brycahta** [comment]
> I think this code might be getting a little too icky.

@jaypipes I'm with ya-- I have the refactor tasks for this package on my radar. I don't disagree with your solution, but to confirm this approach requires refactoring `setResourceReadMany` first, then resolving https://github.com/aws-controllers-k8s/community/issues/1285? If so, would you mind commenting this approach on a refactor issue so we can discuss details?

File: pkg/generate/code/set_resource.go [hunk 3]
```diff
@@ -1566,3 +1605,88 @@ func setResourceForScalar(
...
+// - sourceVarName: resp
+// - outputVarName: elem
+// - indentLevel: 1
+//
+// Sample Output (omit formatting for readability):
+// - opening: "for _, iter0 := range resp.Reservations {
+//	for _, elem := range iter0.Instances {"
+// - closing: "break } break }"
+// - updatedIndentLevel: 3
+func generateForRangeLoops(
```
Comment (brycahta) [inline pkg/generate/code/set_resource.go:1633]: `generateForRangeLoops` wasn't in any designs, but something I wanted to experiment with here. Were there any thoughts from @aws-controllers-k8s/code-generator-maintainer ? This kind of abstraction was somewhat convenient for this solution, but extending this idea to other constructs (i.e. generateIf()) could make the code less clear. Either way-- interested in y'alls thoughts as I'm looking to refactor this package soon™️

File: pkg/generate/code/set_resource.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jaypipes) [reply pkg/generate/code/set_resource.go:128]: ah, ok. I noted you named a similar variable below `wfp` and wondered what was different about this one.

**Review by jaypipes — APPROVED** [review]
OK, from looking at the tests, it seems this does indeed solve the immediate problem at hand with RunInstances/DescribeInstances, so I'm going to merge this and work on a cleaner approach separately as I described in my earlier comment. Thanks @brycahta!

**Pull request merged**